python/revng/model/base.py

```diff
@@ -0,0 +1,170 @@
+                else:
+                    constructor_kwargs[field_name] = Reference.from_type(field_value)
+
+            elif isinstance(field_value, field_spec_type):
+                constructor_kwargs[field_name] = field_value
+
+            elif isinstance(field_value, str) and issubclass(field_spec_type, Enum):
+                enum_instance = field_spec_type(field_value)
+                constructor_kwargs[field_name] = enum_instance
+
+            elif isinstance(field_value, dict) and issubclass(field_spec_type, StructBase):
+                field_instance = field_spec_type.from_dict(**field_value)
+                constructor_kwargs[field_name] = field_instance
+
+            elif isinstance(field_value, str) and hasattr(field_spec_type, "from_string"):
+                field_instance = field_spec_type.from_string(field_value)
+                constructor_kwargs[field_name] = field_instance
+
+            else:
+                raise TypeError(f"Invalid type {type(field_value)} for field {field_name} of {cls.__name__}, was expecting {field_spec_type}")
+
+        instance = cls(**constructor_kwargs)
+        return instance
+
+    @classmethod
+    def from_string(cls, s):
+        raise NotImplementedError(f"from_string not implemented for {cls.__name__}")
+
+    @classmethod
+    def yaml_constructor(cls, loader, node):
+        mapping = loader.construct_mapping(node, deep=True)
+        return cls.from_dict(**mapping)
+
+    @classmethod
+    def yaml_representer(cls, dumper: yaml.dumper.Dumper, instance):
+        mapping_to_dump = {}
+        for field in fields(cls):
+            field_val = instance.__getattribute__(field.name)
+            if _field_is_optional(field) and (field_val is None or field_val == []):
+                continue
+            mapping_to_dump[field.name] = field_val
+
+        return dumper.represent_mapping(f"!{cls.__name__}", mapping_to_dump)
+
+    def __post_init__(self):
+        # Before python 3.10 dataclasses had an annoying limitation regarding inheritance:
+        # if the base class has a field with a default value, *all* the fields of the child classes
+        # need to have a default value too, or dataclasses will raise a TypeError.
+        # Since python 3.10 this limitation can be mostly bypassed by defining the base class
+        # fields as kw_only, but we want to support older python versions.
+        # Hence this workaround, inspired by https://stackoverflow.com/a/53085935
+        for field in fields(self):
+            if self.__getattribute__(field.name) is no_default:
+                raise TypeError(f"__init__ missing 1 required argument: {field.name}")
+
+
+class EnumBase(Enum):
+    @classmethod
+    def yaml_representer(cls, dumper: yaml.dumper.Dumper, instance: Enum):
+        return dumper.represent_str(instance.name)
+
+
+_PointedType = TypeVar("_PointedType")
+_RootType = TypeVar("_RootType")
+
+
+class Reference(Generic[_PointedType, _RootType]):
```

Comment:
This is not really generic as of now. It basically only captures `TypePath`, not all `TupleTreeReference`s.
We'll need to get back to this, but in the meantime, put a bigass note about this.